internal/cms/signed.go

```diff
@@ -0,0 +1,316 @@
+	if signerInfo == nil {
+		return nil, VerificationError{Message: "VerifySigner failed: signer info is required"}
+	}
+
+	if signingCertificate == nil {
+		return nil, VerificationError{Message: "VerifySigner failed: signing certificate is required"}
+	}
+
+	if signerInfo.Version != 1 {
+		// Only IssuerAndSerialNumber is supported currently
+		return nil, VerificationError{Message: fmt.Sprintf("invalid signer info version: only version 1 is supported; got %d", signerInfo.Version)}
+	}
+
+	// user provided signing certificate must match the one in signer info
+	if !bytes.Equal(signingCertificate.RawIssuer, signerInfo.SignerIdentifier.Issuer.FullBytes) || signingCertificate.SerialNumber.Cmp(signerInfo.SignerIdentifier.SerialNumber) != 0 {
+		return nil, SyntaxError{Message: "signing certificate does not match signer info"}
+	}
+
+	return d.verify(signerInfo, signingCertificate, &opts)
+}
+
+// verify verifies the trust in a top-down manner.
+//
+// References:
+//   - RFC 5652 5.4 Message Digest Calculation Process
+//   - RFC 5652 5.6 Signature Verification Process
+func (d *ParsedSignedData) verify(signerInfo *SignerInfo, cert *x509.Certificate, opts *x509.VerifyOptions) (*x509.Certificate, error) {
+	// verify signer certificate
+	certChains, err := cert.Verify(*opts)
+	if err != nil {
+		return cert, VerificationError{Detail: err}
+	}
+
+	// verify signature
+	if err := d.verifySignature(signerInfo, cert); err != nil {
+		return nil, err
+	}
+
+	// verify attribute
+	return cert, d.verifyAttributes(signerInfo, certChains)
+}
+
+// verifySignature verifies the signature with a trusted certificate.
+//
+// References:
+//   - RFC 5652 5.4 Message Digest Calculation Process
+//   - RFC 5652 5.6 Signature Verification Process
+func (d *ParsedSignedData) verifySignature(signerInfo *SignerInfo, cert *x509.Certificate) error {
+	// verify signature
+	algorithm := oid.ToSignatureAlgorithm(
+		signerInfo.DigestAlgorithm.Algorithm,
+		signerInfo.SignatureAlgorithm.Algorithm,
+	)
+	if algorithm == x509.UnknownSignatureAlgorithm {
+		return VerificationError{Message: "unknown signature algorithm"}
+	}
+
+	signed := d.Content
+	if len(signerInfo.SignedAttributes) > 0 {
+		encoded, err := asn1.MarshalWithParams(signerInfo.SignedAttributes, "set")
+		if err != nil {
+			return VerificationError{Message: "invalid signed attributes", Detail: err}
+		}
+		signed = encoded
+	}
+
+	if err := cert.CheckSignature(algorithm, signed, signerInfo.Signature); err != nil {
+		return VerificationError{Detail: err}
+	}
+	return nil
+}
+
+// verifyAttributes verifies the signed attributes.
+//
+// References:
+//   - RFC 5652 5.3 SignerInfo Type
+//   - RFC 5652 5.6 Signature Verification Process
+func (d *ParsedSignedData) verifyAttributes(signerInfo *SignerInfo, chains [][]*x509.Certificate) error {
```

Comment:
Let's rename this function to
```suggestion
func (d *ParsedSignedData) verifySignedAttributes(signerInfo *SignerInfo, chains [][]*x509.Certificate) error {
```

Reply:
Updated.